contracts/Wallet.sol

```diff
@@ -296,58 +282,79 @@ contract Wallet is WalletInterface {
       AdvancedCall memory a = calls[i];
       uint256 callValue = uint256(a.value);
       bytes memory callData = a.data;
+      uint256 callIndex;
 
-      // Note: this check could (should?) be performed prior to execution.
-      if (a.replaceValue.perform) {
-        if (i <= a.replaceValue.callIndex) {
-          revert(_revertReason(36));
+      // Perform low-level call and set return values using result.
+      (bool ok, bytes memory returnData) = a.to.call{value: callValue}(callData);
+      callResults[i] = AdvancedCallReturn({
+          ok: ok,
+          returnData: returnData,
+          callValue: uint96(callValue),
+          callData: callData
+      });
+      if (!ok) {
+        // Exit early - any calls after the first failed call will not execute.
+        rollBack = true;
+        break;
+      }
+
+      for (uint256 j = 0; j < a.replaceValue.length; j++) {
+        callIndex = uint256(a.replaceData[j].callIndex);
```

Comment:
should this be `callIndex = uint256(a.replaceValue[j].callIndex);`?

Reply:
aaaah good catch

Reply:
we might consider writing some tests for the contract!

Reply:
cool that'll be $100k for the audit

Reply:
fixed!